src/main/webapp/app/modules/canvas/visualization/canvas.tsx

```diff
@@ -0,0 +1,279 @@
+  const [visualmetadataList, setvisualmetadata] = useState<IVisualMetadataSet[]>();
+
+  const [filters, setFilters] = useState<string[]>();
+
+  useEffect(() => {
+    if (props.selectedFilter) {
+      const filterList = Object.keys(props.selectedFilter);
+      setFilters(filterList);
+    }
+  }, [props.selectedFilter]);
+
+  const onLayoutChange = _visualmetaList => {
+    props.visualmetadata.visualMetadataSet.map((item, i) => {
+      (item.x = _visualmetaList[i].x),
+        (item.y = _visualmetaList[i].y),
+        (item.h = _visualmetaList[i].h),
+        (item.w = _visualmetaList[i].w),
+        (item.xPosition = _visualmetaList[i].x),
+        (item.yPosition = _visualmetaList[i].y),
+        (item.height = _visualmetaList[i].h),
+        (item.width = _visualmetaList[i].w);
+    });
+  };
+
+  const onResizeStop = (layout, oldItem, newItem, placeholder, e, element) => {
+    const v = VisualMetadataContainerGetOne(oldItem.i);
+    if (v && v.data?.length > 0) {
+      renderVisualization(v, v.data);
+    }
+  };
+
+  const onExchangeMetadataError = data => {
+    const body = JSON.parse(data.body || '{}');
+    // console.log('received error=' + body);
+  };
+
+  const onExchangeMetadata = data => {
+    const metaData = data.body === '' ? { data: [] } : JSON.parse(data.body);
+    if (data.headers.request === 'filters') {
+      const obj = metaData.data[0];
+      let dimensionName = '';
+      for (const i in obj) {
+        dimensionName = i;
+        break;
+      }
+      const retVal = metaData.data.map(function (item) {
+        return {
+          value: item[dimensionName],
+          label: item[dimensionName],
+        };
+      });
+      props.filterData[dimensionName] = retVal;
+    } else {
+      const v = VisualMetadataContainerGetOne(data.headers.queryId);
+      if (v && metaData.data.length > 0) {
+        v.data = metaData.data;
+        $(`.loader-${v.id}`).hide();
+        $(`.dataNotFound-${v.id}`).hide();
+        renderVisualization(v, metaData.data);
+      } else {
+        $(`.dataNotFound-${v.id}`).show();
+        $(`.loader-${v.id}`).hide();
+      }
+    }
+  };
+
+  const renderVisualizationById = item => {
+    if (ValidateFields(item.fields)) {
+      getVisualizationData(item, props.view);
+    } else {
+      $(`.loader-${item.id}`).hide();
+    }
+  };
+
+  const loadVisualization = () => {
+    props.visualmetadata.visualMetadataSet.map((item, i) => {
+      renderVisualizationById(item);
+    });
+  };
+  const connectWeb = () => {
+    connectWebSocket({ token: getToken() }, function (frame) {
+      // console.log(' connected web socket');
+      subscribeWebSocket('/user/exchange/metaData', onExchangeMetadata);
+      subscribeWebSocket('/user/exchange/metaDataError', onExchangeMetadataError);
+      setSocketConnaction(true);
+      VisualMetadataContainerAdd(props.visualmetadata?.visualMetadataSet);
+      loadVisualization();
+    });
+  };
```

Comment:
this should be part of reducer. following advantages
1)you wont need to write connectWebSocket code again in each component. code of duplicacy avoided
2)which ever the component subscribe to this component will receive the data. no need pass the data as props to component or down to compoents

Reply:
done